src/wp-admin/includes/class-theme-upgrader.php

```diff
@@ -566,15 +566,27 @@ public function check_package( $source ) {
 			);
 		}
 
-		// If it's not a child theme, it must have at least an index.php to be legit.
-		if ( empty( $info['Template'] ) && ! file_exists( $working_directory . 'index.php' ) ) {
+		/*
+		 * Parent themes must contain an index file:
+		 * - classic themes require /index.php
+		 * - block themes require /templates/index.html
+		 */
+		if (
+			empty( $info['Template'] ) &&
+			! file_exists( $working_directory . 'index.php' ) &&
+			! file_exists( $working_directory . 'templates/index.html' )
+		) {
 			return new WP_Error(
 				'incompatible_archive_theme_no_index',
 				$this->strings['incompatible_archive'],
 				sprintf(
-					/* translators: %s: index.php */
-					__( 'The theme is missing the %s file.' ),
-					'<code>index.php</code>'
+					/* translators: 1: templates/index.html, 2: index.php, 3: Documentation URL, 4: Template, 5: style.css */
+					__( 'Template is missing. Standalone themes need to have a %1$s or %2$s template file. <a href="%3$s">Child themes</a> need to have a %4$s header in the %5$s stylesheet.' ),
```

Comment:
Thinking this error message is a bit "too much information"? Who is it intended for? :) 
- If it's mostly for users it would be enough to say that the theme seems damaged, and to point them to possible next step (try again, etc.).
- If it's mostly for developers, perhaps it can be a bit more technical, something like "the zip file is either damaged or it doesn't contain a valid WordPress theme as it is missing %1$s and %2$s".

Reply:
> Thinking this error message is a bit "too much information"? Who is it intended for? :)

Are you able to create a follow up ticket for 6.1 to reconsider this string for both uploading and activating themes? 

This ticket is on the 6.0 milestone so needs to use an existing string as the branch is in a hard string freeze.

Reply:
Ah, right, added ti here as this hasn't been committed yet, but it is too late for the string change. Will open a follow-up ticket.